sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+functions will always return `1`.
+
+*Functor kernel*
+```cpp
+using AccTy = cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write,
+                                 cl::sycl::access::target::global_buffer>;
+class Functor {
+public:
+  Functor(int X, AccTy &Acc) : X(X), Acc(Acc) {}
+
+  EXPLICIT_SIMD void operator()() { Acc[0] += X; }
+
+private:
+  int X;
+  AccTy Acc;
+};
+```
+
+*Lambda kernel and function*
+```cpp
+EXPLICIT_SIMD SYCL_EXTERNAL
+void sycl_device_f(cl::sycl::global_ptr<int> ptr, sycl::intel::gpu::simd<float, 8> X) {
+  sycl::intel::gpu::flat_block_write(*ptr.get(), X);
+}
+...
+  Q.submit([&](cl::sycl::handler &Cgh) {
+    auto Acc1 = Buf1.get_access<cl::sycl::access::mode::read>(Cgh);
+    auto Acc2 = Buf2.get_access<cl::sycl::access::mode::read_write>(Cgh);
+
+    Cgh.single_task<class KernelID>([=] () EXPLICIT_SIMD {
+      sycl::intel::gpu::simd<float, 8> Val = sycl::intel::gpu::flat_block_read(Acc1.get_pointer());
+      sycl_device_f(Acc2, Val);
+    });
+  });
+}
+```
+
+## Implementation restrictions
+Current ESP implementation does not support using certain standard SYCL features
+inside explicit SIMD kernels and functions. Most of them will be eventually
+dropped. What's not supported today:
+- Mixing `EXPLICIT_SIMD` kernels with SYCL kernels in a single source
+- Local accessors. Local memory is allocated and accessed via explicit
+device-side API
+- 2D and 3D accessors
+- Constant accessors
+- `sycl::accessor::get_pointer()`. All memory accesses through an accessor are
+done via explicit APIs; e.g. `sycl::intel::gpu::block_store(acc, offset)`
+- Few others (to be documented)
+
+
+## Core Explicit SIMD programming APIs
+
+The DPC++ Explicit SIMD library defines the following classes to enhance the
+expressiveness for explicit SIMD data-parallel programming while enabling
+efficient mapping to SIMD vector operations on Intel GPU architectures.
+
+### SIMD vector class
+
+The ```sycl::intel::gpu::simd``` class is defined as a templated class for a vector
+of fundamental data elements. The fundamental data element type **__Ty** must be
+vectorizable, which can be any of the C++ character/integer/real floating-point types,
+or IEEE-754 half-precision floating-point type. The length of the vector
+is a constant expression, as determined by the template parameter **__N** in a
+given specialization.
+
+Each simd class object is mapped to a consecutive block of general register
+files (GRF).
+
+```cpp
+namespace cm {
+namespace gen {
+template <typename __Ty, int __N> class simd {
+public:
+  using value_type = simd<__Ty, __N, 0>;
+  using element_type = __Ty;
+  using vector_type = __vector_type_t<__Ty, __N>;
+  using type = simd<__Ty, __N>;
+  static constexpr int length = __N;
+
+  // Constructors.
+  constexpr simd();
+  constexpr simd(const simd &other);
+  constexpr simd(simd &&other);
+  constexpr simd(const vector_type &__Val);
+  constexpr simd(std::initializer_list<__Ty> __Ilist) noexcept;
```

Comment:
It's an implementation choice so we can leverage std initializer list syntax for simd vector constructor (e.g., simd<int, 16> v3{0, 2, 4, 6, 1, 3, 5, 7};). These constructors have a simple implementation which are expected to be fully optimized by compiler without throwing.

Reply:
Providing both an overload of initializer-list and the same type itself causes really weird behavior.
E.g.

    simd s1(1,2); //calls next constructor
    simd s2{1,2}; //calls this constructor
This might not be confusing for all users but to everyone using uniform-initialization syntax. Therefore if you want to use this constructor the other one should have a special type (see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es64-use-the-tenotation-for-construction) to avoid this issue.

Also this seems like one of those areas where this simd-type needless differs from std::simd. Why should these constructors be different? Why reinvent the wheel and have all the work of fixing these problems if we could just use the existing solution. Especially if that is anyhow the long-term goal. Adding extra stuff like the select is totally fine. But differ on things which have no apparent advantage and aren't as thought through seems to have only downsides.

> These constructors have a simple implementation which are expected to be fully optimized by compiler without throwing.
And the other constructors (like the copy-constructor) which aren't declared as noexcept, they might throw?